packages/smooth_app/lib/main.dart

```diff
@@ -53,54 +53,44 @@ class _SmoothAppState extends State<SmoothApp> {
   late LocalDatabase _localDatabase;
   late ThemeProvider _themeProvider;
   bool systemDarkmodeOn = false;
-
-  Future<void> _init(BuildContext context) async {
-    _userPreferences = await UserPreferences.getUserPreferences();
-    _productPreferences = ProductPreferences(
-      ProductPreferencesSelection(
-        setImportance: (
-          String attributeId,
-          String importanceId,
-        ) async =>
-            _userPreferences.setImportance(attributeId, importanceId),
-        getImportance: (String attributeId) =>
-            _userPreferences.getImportance(attributeId),
-        notify: () => _productPreferences.notifyListeners(),
-      ),
-    );
-    try {
-      await _productPreferences.loadReferenceFromAssets(
-        DefaultAssetBundle.of(context),
-      );
-    } catch (e) {
-      // this is problematic - we should always be able to load the default
-      debugPrint('Could not load reference files: $e');
-      rethrow;
-    }
-    await _userPreferences.init(_productPreferences);
-    try {
-      _localDatabase = await LocalDatabase.getLocalDatabase();
-    } catch (e) {
-      // this is problematic - we should always be able to init the database
-      debugPrint('Cannot init database: $e');
-      rethrow;
-    }
-    _themeProvider = ThemeProvider(_userPreferences);
-  }
+  late Future<void> _initFuture;
 
   @override
   void initState() {
+    super.initState();
+    _initFuture = _init();
+  }
+
+  Future<void> _init() async {
+    Function debugPrintAndRethrow(String message) => (dynamic error) {
+          debugPrint('$message: $error');
```

Comment:
I'm not sure what the state of the art is: in that case it's not just a lazy "debug print", it might be relevant to actually see that print even in prod conditions (assuming that `debugPrint` prints only in debug mode).
I can see that you did not invent that, I assume it was when @gspencergoog massively upgraded the code recently, but now that I see if that's a typical case where we should use `print`. Or the exception carries enough data and displays it anyway, in that case we don't need `print` or `debugPrint`.

Reply:
`debugPrint` only prints in debug builds.  Usually printing in a release build is bad form: the messages are not seen, and can spam logs on mobile devices. It's better to throw an exception, because then you'll also get a stack trace, and the app will exit. If it's not a fatal error, then it is still better to throw an exception: you can catch it higher up and activate an error reporting mechanism instead.

Reply:
@gspencergoog Right here we're talking about fatal errors, like not being able to open the database or the preferences: the app can't do anything without that.

Coming from Java I was a bit disappointed by flutter/dart's more lenient exception management. I can't remember seeing a flutter app actually crashing - it just logs the exception and sometimes displays an error message in a random `Widget`. In addition to that in our case the exception is thrown in a `Future`.

What would be the best way to handle this particular case: displaying a full screen `Text(errorMessage)` in case an exception is thrown at init time?

@slava-sh I known, this comment obviously goes beyond the initial purpose of this PR. If needed I'll implement @gspencergoog's possible suggestions in another PR, after approving this one.

Reply:
@monsieurtanuki You can use packages like https://pub.dev/packages/sentry_flutter to catch and report errors, and you can also just do it yourself by running the top level app in a zone, and catching errors from that zone.

Assuming the existence of `sender.sendReport`, something like this:

```dart
void main() {
  FlutterError.onError = (FlutterErrorDetails details) {
    sender.sendReport(
      error: details.exception,
      stackTrace: details.stack,
    );
  };

  runZonedGuarded<void>(() {
    // Errors from this call won't be caught by Flutter
    await model.getData();
    // Errors from MyApp constructor won't be caught by FlutterError
    runApp(new MyApp());
  }, onError: (Object exception, StackTrace stackTrace) {
    sender.sendReport(
      error: exception,
      stackTrace: stackTrace,
    );
  });
}
```

Instead of sending a report, you can also recognize the exception and set something on a model to change the widget tree to display a message instead (e.g. a fullscreen page with a message and a "Retry" button).

For a particular future, you can also just use a try/catch block around it to catch an exception that it throws.

Reply:
Created https://github.com/openfoodfacts/smooth-app/issues/480